### PR TITLE
Execute transactions in corresponding resources (also session, plugin)

### PR DESCRIPTION
**Pull request by vogdb**
Example:
While processing incoming transaction with session_id
- go into session
- execute transaction in the session

### CODE REVIEW

**Comment by vogdb** [comment]
@tomaszdurka please look at the idea. I didn't implement it for `Plugin` and `Connection` yet. Another solution would be to make a global available `Transactions` collection. Need you opinion here. Also this solution would require to use WebsocketConnection in composition with Connection instead of inheritance.

**Comment by tomaszdurka** [comment]
I understand we have 3 solutions:

### 1. Use inheritance
Use node inherits utility method. This extends prototype including constructor, allowing usages of `instanceof` etc.

### 2. Use composition
Create instances of needed resources and proxy calls to them


### 3. Trait-like concept
Basically extend prototype. This is the same as Trait concept. You cannot have constructor in trait, simply run a method in the original constructor.

### Proposition

What about:
- do not use Inheritance for all those functionality classes
- use Trait-like for `EventEmitter`, `TransactionGateway`
- use Trait-like or Composition for `WebSocketConnection`

wdyt?

**Comment by vogdb** [comment]
ok. Lets try this Trait-like mechanism. I always avoided it. Also what are name alternatives do you see for TransactionGateway?

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by tomaszdurka** [comment]
Still not sure about this one. I would prefer to stick to normal node classes (EventEmitter).
I understand that my previous proposition about simple extend of prototype (without constructor) is a bit tricky.

What about going back to composition and proxying methods we want (`on`, `emit`)?
Wdyt?

**Comment by vogdb** [comment]
> What about going back to composition and proxying methods we want (`on`, `emit`)?

Back to composition where? Connection uses WebsocketConnection by composition. Other (Plugin, Session, WebsocketConnection) previously extended EventEmitter. Do you want them to refuse from traits? Do not go back to inheritance but use composition. Which means that they have an EventEmitter instance in them and proxy necessary methods to it?

**Comment by vogdb** [comment]
@tomaszdurka wdyt about this?

File: src/connection.js [hunk 1]
```diff
@@ -18,28 +19,55 @@ var Session = require('./session');
...
+  /** @type {Object} */
   this._sessions = {};
+
+  /** @type {WebsocketConnection} */
+  this._websocketConnection = new WebsocketConnection();
+  this._installWebsocketListeners();
 }
 
-util.inherits(Connection, WebsocketConnection);
+Helpers.extend(Connection.prototype, TEventEmitter, TTransactionGateway);
```
Comment (tomaszdurka) [inline src/connection.js]: Hmm maybe...
```
Helpers.extend.call(Connection.prototype, EventEmitter, TTransactionGateway)
```
:panda_face:

Comment (vogdb) [inline src/connection.js]: This is a very fragile construction. You can not simply copy EventEmitter into Connection.prototype and expect that Connection will have emitter's API. EventEmitter is also a private state which can't be copied by simple extend. The approach that you can use is `Helpers.extend(Connection.prototype, EventEmitter.prototype, TTransactionGateway);` but also you need to call for EventEmitter's constructor in Connection's constructor.

```js
function Connection(id, options) {
  EventEmitter.call(this);
```
which is not a Trait. It's a partial multiple inheritance.

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by tomaszdurka** [comment]
lgtm

**Pull request merged**